s3torchconnector/tst/e2e/dcp/test_e2e_s3_file_system.py

```diff
@@ -4,19 +4,315 @@
+        "tensor1": torch.randn(10, 10),
+        "tensor2": torch.randn(5, 5),
+        "scalar": torch.tensor(3.14),
+    }
+    mp.spawn(
+        run,
+        args=(
+            world_size,
+            thread_count,
+            region,
+            s3_path_s3storagewriter,
+            s3_path_fsspec,
+            test_data,
+        ),
+        nprocs=world_size,
+        join=True,
+    )
+
+    load_data(
+        region,
+        s3_path_s3storagewriter,
+        s3_path_fsspec,
+        test_data,
+        world_size,
+        thread_count,
+    )
+
+
+def dcp_save(data, writer):
+    dcp.save(
+        data,
+        storage_writer=writer,
+    )
+
+
+def dcp_load(loaded_data, reader):
+    dcp.load(
+        loaded_data,
+        storage_reader=reader,
+    )
+
+
+def load_data(
+    region, s3_path_s3storagewriter, s3_path_fsspec, test_data, world_size, thread_count
+):
+    s3_client = S3Client(region=region)
+    bucket, key = parse_s3_uri(s3_path_s3storagewriter)
+    list_result_s3storagewriter = list(s3_client.list_objects(bucket, f"{key}/"))
+
+    # Compare length
+    assert list_result_s3storagewriter is not None
+    assert (
+        len(list_result_s3storagewriter[0].object_info) == world_size * thread_count + 1
+    )
+
+    # Load using S3StorageReader
+    loaded_data_s3storagereader = {}
+    dcp_load(
+        loaded_data_s3storagereader,
+        S3StorageReader(
+            region,
+            s3_path_s3storagewriter,
+        ),
+    )
+
+    # Load using FsspecReader
+    loaded_data_fsspec = {}
+    dcp_load(
+        loaded_data_fsspec,
+        FsspecReader(
+            s3_path_fsspec,
+        ),
+    )
+
+    # Compare loaded data
+    assert set(loaded_data_s3storagereader.keys()) == set(
+        loaded_data_fsspec.keys()
+    ), "Loaded data keys do not match"
+
+    for key in loaded_data_s3storagereader.keys():
+        assert torch.allclose(
+            loaded_data_s3storagereader[key], loaded_data_fsspec[key]
+        ), f"Loaded tensors for key '{key}' do not match"
+        assert torch.allclose(
+            loaded_data_s3storagereader[key], test_data[key]
+        ), f"S3StorageReader: Loaded tensor for key '{key}' does not match original"
+        assert torch.allclose(
+            loaded_data_fsspec[key], test_data[key]
+        ), f"FsspecReader: Loaded tensor for key '{key}' does not match original"
+
+    print(
+        "Test passed: Both implementations saved and loaded data correctly, and the contents match."
+    )
+
+
+def test_dcp_when_multi_process_single_thread(checkpoint_directory):
+    multi_process_dcp_save_load(6, 1, checkpoint_directory)
+
+
+def test_dcp_when_multi_process_multiple_threads(checkpoint_directory):
+    multi_process_dcp_save_load(6, 4, checkpoint_directory)
+
+
+def test_dcp_save_non_existing_s3_uri(checkpoint_directory):
+    t1 = torch.randn(10)
+    region = checkpoint_directory.region
+    non_existing_s3_uri = "s3://non-existing-bucket/checkpoint"
```

Comment:
Will that bucket name have meaning for S3 Express?

Reply:
No, will change this

Reply:
Actually if you see the s3_uri in checkpoint directory https://github.com/awslabs/s3-connector-for-pytorch/blob/1f8fdd049c52f6d99450fd9ebc9f6a576aed5a58/s3torchconnector/tst/e2e/conftest.py#L42 which is used in previous happy tests and have successfully got executed in s3 express tests, that is also defined in this format. So, on our end we don't take case of creating proper uri for s3 express, it is done finally on the crt end.